flight/PiOS/STM32F10x/startup.c

```diff
@@ -0,0 +1,199 @@
+ * @addtogroup PIOS PIOS Core hardware abstraction layer
+ * @{
+ *
+ * @file       startup.c
+ * @author     dRonin, http://dRonin.org/, Copyright (C) 2016
+ * @author     The OpenPilot Team, http://www.openpilot.org Copyright (C) 2012.
+ * @author     Tau Labs, http://taulabs.org, Copyright (C) 2014
+ * @brief      C based startup of F4
+ * @see        The GNU Public License (GPL) Version 3
+ *
+ *****************************************************************************/
+/*
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
+ * or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License
+ * for more details.
+ *
+ * You should have received a copy of the GNU General Public License along
+ * with this program; if not, write to the Free Software Foundation, Inc.,
+ * 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
+ *
+ * Additional note on redistribution: The copyright and license notices above
+ * must be maintained in each individual source file that is a derivative work
+ * of this source file; otherwise redistribution is prohibited.
+ */
+
+
+#include <string.h>
+#include <stm32f10x.h>
+
+/* prototype for main() that tells us not to worry about it possibly returning */
+extern int		main(void) __attribute__((noreturn));
+
+/* prototype our _main() to avoid prolog/epilog insertion and other related junk */
+void 			_main(void) __attribute__((noreturn, naked, no_instrument_function));
+
+/* stack_change() to reclaim init stack for heap */
+void 			stack_change(void) __attribute__((noreturn, naked, no_instrument_function));
+
+/** default handler for CPU exceptions */
+static void		default_cpu_handler(void) __attribute__((noreturn, naked, no_instrument_function));
+
+extern void 	PIOS_heap_increase_size(size_t bytes);
+
+/** BSS symbols XXX should have a header that defines all of these */
+extern char		_sbss, _ebss;
+
+/** DATA symbols XXX should have a header that defines all of these */
+extern char		_sidata, _sdata, _edata;
+
+/** The IRQ and init stacks */
+char			init_stack[INIT_STACK_SIZE] __attribute__((section(".initstack")));
+char			irq_stack[IRQ_STACK_SIZE] __attribute__((section(".irqstack")));
+
+/** exception handler */
+typedef const void	(vector)(void);
+
+/** CortexM3 CPU vectors */
+struct cm3_vectors {
+	void	*initial_stack;
+	vector	*entry;
+	vector	*vectors[14];
+};
+
+/**
+ * Initial startup code.
+ */
+void
+_main(void)
+{
+	// load the stack base for the current stack before we attempt to branch to any function
+	// that might bounds-check the stack
+	asm volatile ("mov r10, %0" : : "r" (&irq_stack[0]) : );
```

Comment:
We don't, but there's no reason to not do this here.

(I'm thinking about building stack enforcement using R10 to replace the rtos-enforcey-stuff we took out)